src/NITTA/FrontEnds/XMILE/XMILEFrontend.hs

```diff
@@ -0,0 +1,261 @@
+            put x{algNextFreeNameIndex = HM.insert name (index + 1) algNextFreeNameIndex}
+            return $ getGraphName name index
+
+        getAllOutGraphNodes name = do
+            XMILEAlgBuilder{algUsagesCount} <- get
+            let usages = fromMaybe (error $ "name not found : " <> T.unpack name) $ HM.lookup name algUsagesCount
+            return $ map (\num -> getGraphName name num) [0 .. usages]
+
+        getDefaultValue name = do
+            XMILEAlgBuilder{algDefaultValues} <- get
+            return $ fromMaybe (error $ "name not found :" <> T.unpack name) $ HM.lookup name algDefaultValues
+
+        getGraphName name index = name <> T.pack "#" <> showText index
+
+getDefaultValuesAndUsages algBuilder = do
+    mapM_ processStock $ xcStocks algBuilder
+    where
+        nameToEquationMap = flip execState HM.empty $ do
+            mapM_ (addToMap . (\a -> (xaName a, xaEquation a))) (xcAuxs algBuilder)
+            mapM_ (addToMap . (\a -> (xfName a, xfEquation a))) (xcFlows algBuilder)
+            mapM (addToMap . (\a -> (xsName a, xsEquation a))) (xcStocks algBuilder)
+            where
+                addToMap (name, eqn) = do
+                    hm <- get
+                    put $ HM.insert name eqn hm
+        processStock XMILEStock{xsEquation, xsName, xsInflow, xsOutflow} = do
+            alg@XMILEAlgBuilder{algDefaultValues} <- get
+            let val = calculateDefaultValue algDefaultValues xsEquation
+            put alg{algDefaultValues = HM.insert xsName val algDefaultValues}
+            processFlow xsInflow
+            processFlow xsOutflow
+
+            return ()
+            where
+                processFlow Nothing = do return ()
+                processFlow (Just name) = do
+                    addUsages name
+                    addUsages xsName
+                    addDtUsagesIfNeeded $ xssDt $ xcSimSpecs algBuilder
+                    processEquation $ xfEquation $ findFlow name
+                    where
+                        addDtUsagesIfNeeded 1 = do return ()
+                        addDtUsagesIfNeeded _ = do
+                            addUsages deltaTimeVarName
+
+                findFlow name =
+                    fromMaybe
+                        (error $ "cannot find expected flow flow with name " <> T.unpack name <> show (xcFlows algBuilder))
+                        (L.find (\XMILEFlow{xfName} -> xfName == name) $ xcFlows algBuilder)
+
+                processEquation (Val _) = do return ()
+                processEquation (Duo _ expr expl) = do
+                    processEquation expr
+                    processEquation expl
+                processEquation (Var name) = do
+                    addUsages $ T.pack name
+                    addDefaultValueIfNeeded
+                    where
+                        addDefaultValueIfNeeded = do
+                            x@XMILEAlgBuilder{algDefaultValues} <- get
```

Comment:
`x` is not a very conventional name for that builder state. Use `st` and `state`.